### PR TITLE
Add live migration of system VMs (KVM)

### PR DESCRIPTION
**Pull request by BryanMLima**
### Description

Currently, there is no way to migrate running system VMs, using the KVM hypervisor, with volumes; this means, we cannot move their volumes to a new storage pool in runtime. This PR aims to add this feature by altering the behavior of the `migrateVirtualMachineWithVolume` API to support system VMs. Furthermore, the `listVolumes` API was changed to also return system VMs root volumes, to provide easy access to the system VMs volume's UUID to the root admins. For this, a new parameter called `listsystemvms` was added to the `listVolumesCmd`. Moreover, it was added unit test for `MigrateVirtualMachineWithVolumeCmd` class.

Furthermore, the changes made by this feature required some tweaking in the UI as the system VM's volumes were now available. When these volumes were accessed, the path led to `/vm/<vm-uuid>`. However, ACS has a different path to VRs and system VMs resulting in a 404 page. This bug was fixed with the new changes; moreover, it was added a Volumes tab to system VMs and VRs as well.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):

VR volume's tab:  
image

System volume's tab:
image

### How Has This Been Tested?
This was tested on a local lab through CloudMonkey, using the command `migrate virtualmachinewithvolume` with the UUID of a system VM and migrating its volume to another storage pool. The command ran as expected. Furthermore, it was added unit test to verify the API behavior.

For the UI, I tried to access the volumes of system VMs and VRs as well as the volumes of user VMs to check if nothing was broken. I did not find any errors.

### CODE REVIEW

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java
Comment (RodrigoDLopez) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java:196]:
```suggestion
    private void setResponseBasedOnVmType(VirtualMachine userVm, VirtualMachine migratedVm) {
```

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java
Comment (BryanMLima) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java]: You are correct, I made the change.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java:159]: how about the `null` check that used to be there? in `UserVmManagerImpl.getVM(long)` I see no sign of `null`-safety.

Comment (BryanMLima) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java:159]: As I was adjusting the error messages, I could not verify this one. Furthermore, I verified that the method `ParamProcessWorker#translateUuidToInternalId` already makes this verification. If the user puts an nonexistent UUID, or and UUID from another entity, the `translateUuidToInternalId` method already throws an exception, for these reason I removed this `null` check.

**Review by JoaoJandre — APPROVED** [review]
CLGTM

**Comment by DaanHoogland** [comment]
@GabrielBrascher @RodrigoDLopez are your concerns all met?

**Review by DaanHoogland — APPROVED** [review]
cltgm

**Comment by RodrigoDLopez** [comment]
> @GabrielBrascher @RodrigoDLopez are your concerns all met?

Yes, I do agree with this proposal. Before I approved this one, I ran some manual tests in a local lab, and everything was good, I think we are good to go.

**Comment by DaanHoogland** [comment]
> > @GabrielBrascher @RodrigoDLopez are your concerns all met?
> 
> Yes, I do agree with this proposal. Before I approved this one, I ran some manual tests in a local lab, and everything was good, I think we are good to go.

ah, your approval status is still ¨left review comments¨. I understand you have tested this. Let's wait for @GabrielBrascher 's reply and then merge.

**Comment by DaanHoogland** [comment]
amend: let's run smoke tests as well;
@blueorangutan package

**Review by RodrigoDLopez — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@Pearl1594  @rohityadavcloud  could you shine your lights on this, please?

**Comment by BryanMLima** [comment]
Hey @DaanHoogland, could you take a look on this one again? I think it already has the required number of approvals.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java
Comment (GutoVeronezi) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java:156]:
```suggestion
            throw new InvalidParameterValueException(String.format("Either %s or %s must be passed for migrating the VM.", ApiConstants.HOST_ID, ApiConstants.MIGRATE_TO));
```

File: ui/src/views/compute/InstanceTab.vue
Comment (GutoVeronezi) [inline ui/src/views/compute/InstanceTab.vue:37]:
```suggestion
        <volumes-tab :resource="vm" :items="volumes" :loading="loading" />
```

**Comment by GutoVeronezi** [comment]
@BryanMLima, can you check Travis errors?

**Comment by DaanHoogland** [comment]
@Pearl1594 I dare not merge this without your permission ;)

**Comment by BryanMLima** [comment]
Hey @DaanHoogland, I made some changes in the core behavior of this PR. I added a new parameter for the `listVolumes` API to specify when to list system VMs volumes. Therefore, the default behavior is not changed, as it were with the past changes, and the integration tests passed smoothly. If you could review again, I appreciate.

**Review by shwstppr — COMMENTED** [review]
Title of the PR is confusing for me as live migration for system VMs is already supported. May it should read - support migrate with volume for system VMs. Also, this may need testing both UI and API

**Comment by DaanHoogland** [comment]
@BryanMLima can you respond to / apply @shwstppr's comments?

File: api/src/main/java/com/cloud/vm/UserVmService.java
Comment (BryanMLima) [reply api/src/main/java/com/cloud/vm/UserVmService.java:428]: It does, however, to make this happen it caused a circular dependency problem, and to solve it, it would need to refactor a lot of classes to fix the issue. The refactor could be interesting in another PR, but is out of the scope of this one.

**Comment by BryanMLima** [comment]
> Title of the PR is confusing for me as live migration for system VMs is already supported. May it should read - support migrate with volume for system VMs. Also, this may need testing both UI and API

I agree, I forgot to mentioned that these changes are to support live migration of system VMs for KVM. I will make some changes in this PR description to make it clear.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java
Comment (BryanMLima) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/MigrateVirtualMachineWithVolumeCmd.java]: While testing in my local lab, I did not find any problems, but more testing is always better.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by BryanMLima** [comment]
@DaanHoogland @GutoVeronezi can we proceed with the merge on this PR. The Travis CI passed after retrying the failed job.

**Comment by DaanHoogland** [comment]
> @DaanHoogland @GutoVeronezi can we proceed with the merge on this PR. The Travis CI passed after retrying the failed job.

yes @BryanMLima , as soon as @shwstppr confirms his comments are addressed.

File: api/src/main/java/com/cloud/vm/UserVmService.java
Comment (shwstppr) [reply api/src/main/java/com/cloud/vm/UserVmService.java:428]: okay @BryanMLima

**Comment by shwstppr** [comment]
@DaanHoogland cc @BryanMLima my comments have been addressed

**Comment by DaanHoogland** [comment]
thanks for your patience and persistence @BryanMLima

**Pull request merged**
